README.md
# node-canvas-text

Draws your string on a _canvas_, fit inside of a rectangle. I had to make this, because _measureText()_ from node-canvas is unpredictable, and ignores font selectors besides font-size and font-family.

## Requirements:

- [node-canvas](https://github.com/Automattic/node-canvas) to draw onto
- [opentype.js](https://github.com/nodebox/opentype.js/blob/master/README.md) to load OpenType fonts

## Installation

`npm install node-canvas-text canvas opentype.js --save`

## Parameters

This module exports a single function with signature:

1. context from node-canvas
2. a string to draw
3. font object
4. bounding rectangle `{ x, y, width, height }`
5. options `{ minSize, maxSize, granularity, hAlign, vAlign, fitMethod, drawRect }`

### Options

- **minSize**: minimum font size `float`
- **maxSize**: maximum font size `float`
- **granularity**: a step, in which to scale font size `float`
- **hAlign**: horizontal text alignment `'left' | 'center' | 'right'`
- **vAlign**: vertical text alignment `'top' | 'center' | 'bottom'`
- **fitMethod**: `'baseline' | 'box'`
- **drawRect**: draw the bounding rectangle `'true' | 'false'`
- **textFillStyle**: fill style for text `string`
- **rectFillStyle**: fill style for rectangle `string`
- **rectFillOnlyText**: fill only the exact resulting text rectangle, not the bounding one `'true' | 'false'`
- **textPadding**: text padding `float`
- **fillPadding**: fill padding `float`

#### Defaults

```javascript
{
    minSize: 10,
    maxSize: 200,
    granularity: 1,
    hAlign: 'left',
    vAlign: 'bottom',
    fitMethod: 'box',
    textFillStyle: '#000',
    rectFillStyle: '#fff',
    rectFillOnlyText: false,
    textPadding: 0,
    fillPadding: 0,
    drawRect: false
}
```

### Fit method: box vs baseline

![fitMethod: box](http://i.imgur.com/wuLdnPs.jpg)
![fitMethod: baseline](http://i.imgur.com/oxJQvYZ.jpg)

## Example

```javascript
const canvasText = require('node-canvas-text');
const opentype = require('opentype.js');
const Canvas = require('canvas');

let canvas = new Canvas.canvas(imgWidth, imgHeight);
let ctx = canvas.getContext('2d');

// Load OpenType fonts from files
let titleFont = opentype.loadSync(__dirname + '/fonts/PTN57F.ttf');
let priceFont = opentype.loadSync(__dirname + '/fonts/PTC75F.ttf');
let barcodeFont = opentype.loadSync(__dirname + '/fonts/code128.ttf');

// Strings to draw
let titleString = 'A string, but not too long',
  priceString = '200',
  barcodeString = '54490000052117';

// Calculate bounding rectangles
let headerRect = {
  x: 0,
  y: 0,
  width: canvas.width,
  height: canvas.height / 3.5
};

let priceRect = {
  x: canvas.width / 2,
  y: headerRect.height,
  width: canvas.width / 2,
  height: canvas.height - headerRect.height
};

let barcodeRect = {
  x: 0,
  y: headerRect.height + priceRect.height / 2,
  width: canvas.width - priceRect.width,
  height: priceRect.height / 2
};

// Draw
let drawRect = true;

canvasText.drawText(ctx, titleString, titleFont, headerRect, {
  minSize: 5,
  maxSize: 100,
  vAlign: 'bottom',
  hAlign: 'left',
  fitMethod: 'box',
  drawRect: drawRect
});

canvasText.drawText(ctx, priceString, priceFont, priceRect, {
  minSize: 5,
  maxSize: 200,
  hAlign: 'right',
  vAlign: 'bottom',
  fitMethod: 'box',
  drawRect: drawRect
});

canvasText.drawText(ctx, barcodeString, barcodeFont, barcodeRect, {
  minSize: 5,
  maxSize: 200,
  hAlign: 'center',
  vAlign: 'center',
  fitMethod: 'box',
  drawRect: drawRect
});
```


CHANGELOG.md
# 1.0.3 / 2019-02-23

- .default() is now .drawText()

# 1.0.2 / 2016-03-01

- **textFillStyle** and **rectFillStyle** now behave properly
- The default for **rectFillStyle** is now _transparent_


index.dev.js
const measureText = (text, font, fontSize, method = 'box') => {
  let ascent = 0,
    descent = 0,
    width = 0,
    scale = (1 / font.unitsPerEm) * fontSize,
    glyphs = font.stringToGlyphs(text);

  for (var i = 0; i < glyphs.length; i++) {
    let glyph = glyphs[i];
    if (glyph.advanceWidth) {
      width += glyph.advanceWidth * scale;
    }
    if (i < glyphs.length - 1) {
      let kerningValue = font.getKerningValue(glyph, glyphs[i + 1]);
      width += kerningValue * scale;
    }

    let { yMin, yMax } = glyph.getMetrics();

    ascent = Math.max(ascent, yMax);
    descent = Math.min(descent, yMin);
  }

  return {
    width: width,
    height:
      method == 'box'
        ? Math.abs(ascent) * scale + Math.abs(descent) * scale
        : Math.abs(ascent) * scale,
    actualBoundingBoxAscent: ascent * scale,
    actualBoundingBoxDescent: descent * scale,
    fontBoundingBoxAscent: font.ascender * scale,
    fontBoundingBoxDescent: font.descender * scale
  };
};

const padRectangle = (rectangle, padding) => {
  return {
    x: rectangle.x - padding,
    y: rectangle.y - padding,
    width: rectangle.width + padding * 2,
    height: rectangle.height + padding * 2
  };
};

exports.drawText = (ctx, text, fontObject, _rectangle = {}, _options = {}) => {
  let paddedRect = {
    ...{
      x: 0,
      y: 0,
      width: 100,
      height: 100
    },
    ..._rectangle
  };

  let options = {
    ...{
      minSize: 10,
      maxSize: 200,
      granularity: 1,
      hAlign: 'left',
      vAlign: 'bottom',
      fitMethod: 'box',
      textFillStyle: '#000',
      rectFillStyle: 'transparent',
      rectFillOnlyText: false,
      textPadding: 0,
      fillPadding: 0,
      drawRect: false
    },
    ..._options
  };

  if (typeof text != 'string') throw 'Missing string parameter';
  if (typeof fontObject != 'object') throw 'Missing fontObject parameter';
  if (typeof ctx != 'object') throw 'Missing ctx parameter';
  if (options.minSize > options.maxSize)
    throw 'Min font size can not be larger than max font size';

  let originalRect = paddedRect;
  paddedRect = padRectangle(paddedRect, options.textPadding);

  ctx.save();

  let fontSize = options.maxSize;
  let textMetrics = measureText(text, fontObject, fontSize, options.fitMethod);
  let textWidth = textMetrics.width;
  let textHeight = textMetrics.height;

  while (
    (textWidth > paddedRect.width || textHeight > paddedRect.height) &&
    fontSize >= options.minSize
  ) {
    fontSize = fontSize - options.granularity;
    textMetrics = measureText(text, fontObject, fontSize, options.fitMethod);
    textWidth = textMetrics.width;
    textHeight = textMetrics.height;
  }

  // Calculate text coordinates based on options
  let xPos = paddedRect.x;
  let yPos =
    options.fitMethod == 'box'
      ? paddedRect.y +
        paddedRect.height -
        Math.abs(textMetrics.actualBoundingBoxDescent)
      : paddedRect.y + paddedRect.height;

  switch (options.hAlign) {
    case 'right':
      xPos = xPos + paddedRect.width - textWidth;
      break;
    case 'center':
    case 'middle':
      xPos = xPos + paddedRect.width / 2 - textWidth / 2;
      break;
    case 'left':
      break;
    default:
      throw 'Invalid options.hAlign parameter: ' + options.hAlign;
      break;
  }

  switch (options.vAlign) {
    case 'top':
      yPos = yPos - paddedRect.height + textHeight;
      break;
    case 'center':
    case 'middle':
      yPos = yPos + textHeight / 2 - paddedRect.height / 2;
      break;
    case 'bottom':
    case 'baseline':
      break;
    default:
      throw 'Invalid options.vAlign parameter: ' + options.vAlign;
      break;
  }

  ctx.fillStyle = 'transparent';

  // Draw fill rectangle if needed
  if (options.rectFillStyle != 'transparent') {
    let fillRect = options.rectFillOnlyText
      ? {
          x: xPos,
          y: yPos - textHeight,
          width: textWidth,
          height: textHeight
        }
      : originalRect;

    fillRect = padRectangle(fillRect, options.fillPadding);

    ctx.fillStyle = options.rectFillStyle;
    ctx.fillRect(fillRect.x, fillRect.y, fillRect.width, fillRect.height);
    ctx.fillStyle = 'transparent';
  }

  // Draw text
  let fontPath = fontObject.getPath(text, xPos, yPos, fontSize);
  fontPath.fill = options.textFillStyle;
  fontPath.draw(ctx);

  // Draw bounding rectangle
  if (options.drawRect) {
    // TODO: Figure out how to not stroke the text itself, just the rectangle
    ctx.save();
    ctx.strokeStyle = 'red';
    ctx.rect(paddedRect.x, paddedRect.y, paddedRect.width, paddedRect.height);
    ctx.stroke();
    ctx.strokeStyle = 'transparent';
    ctx.restore();
  }

  ctx.restore();
};


index.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var measureText = function measureText(text, font, fontSize) {
  var method = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 'box';
  var ascent = 0,
      descent = 0,
      width = 0,
      scale = 1 / font.unitsPerEm * fontSize,
      glyphs = font.stringToGlyphs(text);

  for (var i = 0; i < glyphs.length; i++) {
    var glyph = glyphs[i];

    if (glyph.advanceWidth) {
      width += glyph.advanceWidth * scale;
    }

    if (i < glyphs.length - 1) {
      var kerningValue = font.getKerningValue(glyph, glyphs[i + 1]);
      width += kerningValue * scale;
    }

    var _glyph$getMetrics = glyph.getMetrics(),
        yMin = _glyph$getMetrics.yMin,
        yMax = _glyph$getMetrics.yMax;

    ascent = Math.max(ascent, yMax);
    descent = Math.min(descent, yMin);
  }

  return {
    width: width,
    height: method == 'box' ? Math.abs(ascent) * scale + Math.abs(descent) * scale : Math.abs(ascent) * scale,
    actualBoundingBoxAscent: ascent * scale,
    actualBoundingBoxDescent: descent * scale,
    fontBoundingBoxAscent: font.ascender * scale,
    fontBoundingBoxDescent: font.descender * scale
  };
};

var padRectangle = function padRectangle(rectangle, padding) {
  return {
    x: rectangle.x - padding,
    y: rectangle.y - padding,
    width: rectangle.width + padding * 2,
    height: rectangle.height + padding * 2
  };
};

exports.drawText = function (ctx, text, fontObject) {
  var _rectangle = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};

  var _options = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : {};

  var paddedRect = _objectSpread({}, {
    x: 0,
    y: 0,
    width: 100,
    height: 100
  }, _rectangle);

  var options = _objectSpread({}, {
    minSize: 10,
    maxSize: 200,
    granularity: 1,
    hAlign: 'left',
    vAlign: 'bottom',
    fitMethod: 'box',
    textFillStyle: '#000',
    rectFillStyle: 'transparent',
    rectFillOnlyText: false,
    textPadding: 0,
    fillPadding: 0,
    drawRect: false
  }, _options);

  if (typeof text != 'string') throw 'Missing string parameter';
  if (_typeof(fontObject) != 'object') throw 'Missing fontObject parameter';
  if (_typeof(ctx) != 'object') throw 'Missing ctx parameter';
  if (options.minSize > options.maxSize) throw 'Min font size can not be larger than max font size';
  var originalRect = paddedRect;
  paddedRect = padRectangle(paddedRect, options.textPadding);
  ctx.save();
  var fontSize = options.maxSize;
  var textMetrics = measureText(text, fontObject, fontSize, options.fitMethod);
  var textWidth = textMetrics.width;
  var textHeight = textMetrics.height;

  while ((textWidth > paddedRect.width || textHeight > paddedRect.height) && fontSize >= options.minSize) {
    fontSize = fontSize - options.granularity;
    textMetrics = measureText(text, fontObject, fontSize, options.fitMethod);
    textWidth = textMetrics.width;
    textHeight = textMetrics.height;
  } // Calculate text coordinates based on options


  var xPos = paddedRect.x;
  var yPos = options.fitMethod == 'box' ? paddedRect.y + paddedRect.height - Math.abs(textMetrics.actualBoundingBoxDescent) : paddedRect.y + paddedRect.height;

  switch (options.hAlign) {
    case 'right':
      xPos = xPos + paddedRect.width - textWidth;
      break;

    case 'center':
    case 'middle':
      xPos = xPos + paddedRect.width / 2 - textWidth / 2;
      break;

    case 'left':
      break;

    default:
      throw 'Invalid options.hAlign parameter: ' + options.hAlign;
      break;
  }

  switch (options.vAlign) {
    case 'top':
      yPos = yPos - paddedRect.height + textHeight;
      break;

    case 'center':
    case 'middle':
      yPos = yPos + textHeight / 2 - paddedRect.height / 2;
      break;

    case 'bottom':
    case 'baseline':
      break;

    default:
      throw 'Invalid options.vAlign parameter: ' + options.vAlign;
      break;
  }

  ctx.fillStyle = 'transparent'; // Draw fill rectangle if needed

  if (options.rectFillStyle != 'transparent') {
    var fillRect = options.rectFillOnlyText ? {
      x: xPos,
      y: yPos - textHeight,
      width: textWidth,
      height: textHeight
    } : originalRect;
    fillRect = padRectangle(fillRect, options.fillPadding);
    ctx.fillStyle = options.rectFillStyle;
    ctx.fillRect(fillRect.x, fillRect.y, fillRect.width, fillRect.height);
    ctx.fillStyle = 'transparent';
  } // Draw text


  var fontPath = fontObject.getPath(text, xPos, yPos, fontSize);
  fontPath.fill = options.textFillStyle;
  fontPath.draw(ctx); // Draw bounding rectangle

  if (options.drawRect) {
    // TODO: Figure out how to not stroke the text itself, just the rectangle
    ctx.save();
    ctx.strokeStyle = 'red';
    ctx.rect(paddedRect.x, paddedRect.y, paddedRect.width, paddedRect.height);
    ctx.stroke();
    ctx.strokeStyle = 'transparent';
    ctx.restore();
  }

  ctx.restore();
};
